### PR TITLE
node: allow websocket and HTTP on the same port

### PR DESCRIPTION
**Pull request by renaynay**
This change makes it possible to run geth with HTTP and WebSocket RPC
on the same TCP port:

    geth --rpc --rpcport 8545 --ws --wsport 8545

**TODOs:** 
- [x] functional implementation of handling for websocket and http on same port
- [x] handler stack needs to be pulled out of rpc pkg and created in node pkg instead to avoid use of deprecated `rpc.NewHTTPServer` method
- [x] Document new methods / where necessary
- [x] **TESTS!**
- [x] general clean-up of code
    - [x] add logging for debug / info
    - [x] make sure tests are functioning and no code is commented out for sake of functionality
    - [x] make sure no other packages are impacted by changes, and if they are, handle for that.
- [x] move commits associated with comment clean-up into separate PR
- [x] nice-haves: clean up websocket portion so that it matches http portion (meaning pull out handler stack into node pkg to avoid use of ws deprecated newWsServer method)

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -359,26 +364,53 @@ func (n *Node) stopIPC() {
...
+func (n *Node) startHTTP(endpoint string, apis []rpc.API, modules []string, cors []string, vhosts []string, timeouts rpc.HTTPTimeouts, wsOrigins []string) error {
 	// Short circuit if the HTTP endpoint isn't being exposed
 	if endpoint == "" {
 		return nil
 	}
-	listener, handler, err := rpc.StartHTTPEndpoint(endpoint, apis, modules, cors, vhosts, timeouts)
+
+	srv := rpc.NewServer()
+
+	err := RegisterApisFromWhitelist(apis, modules, srv)
```
Comment (holiman) [inline node/node.go]: don't forget to handle this guy -- you're overwriting it on L385

File: node/node.go [hunk 2]
```diff
@@ -359,26 +364,53 @@ func (n *Node) stopIPC() {
...
+
+	var ws http.Handler
+	if n.httpEndpoint == n.wsEndpoint {
+		ws = srv.WebsocketHandler(wsOrigins)
+	}
+
```
Comment (holiman) [inline node/node.go]: @karalabe hates whitespace

File: cmd/clef/main.go [hunk 3]
```diff
@@ -539,9 +539,17 @@ func signer(c *cli.Context) error {
 		vhosts := splitAndTrim(c.GlobalString(utils.RPCVirtualHostsFlag.Name))
 		cors := splitAndTrim(c.GlobalString(utils.RPCCORSDomainFlag.Name))
 
-		// start http server
+		srv := rpc.NewServer()
+
+		err := node.RegisterApisFromWhitelist(rpcAPI, []string{"account"}, srv, false)
+		if err != nil {
+			utils.Fatalf("Could not register API: %w", err) // TODO should this be a fatal failure?
```
Comment (holiman) [inline cmd/clef/main.go]:
```suggestion
			utils.Fatalf("Could not register API: %w", err)
```
yes, that's correct

File: cmd/geth/retesteth.go [hunk 4]
```diff
@@ -888,14 +888,22 @@ func retesteth(ctx *cli.Context) error {
 	vhosts := splitAndTrim(ctx.GlobalString(utils.RPCVirtualHostsFlag.Name))
 	cors := splitAndTrim(ctx.GlobalString(utils.RPCCORSDomainFlag.Name))
 
+	srv := rpc.NewServer()
+
+	err := node.RegisterApisFromWhitelist(rpcAPI, []string{"test", "eth", "debug", "web3"}, srv, false)
+	if err != nil {
+		utils.Fatalf("Could not register RPC apis: %w", err) // TODO should this be a fatal error?
```
Comment (holiman) [inline cmd/geth/retesteth.go]:
```suggestion
		utils.Fatalf("Could not register RPC apis: %w", err)
```

File: node/node.go [hunk 5]
```diff
@@ -664,3 +689,26 @@ func (n *Node) apis() []rpc.API {
...
+	for _, module := range modules {
+		whitelist[module] = true
+	}
+	// Register all the APIs exposed by the services
+	for _, api := range apis {
+		if exposeAll || whitelist[api.Namespace] || (len(whitelist) == 0 && api.Public) {
+			if err := srv.RegisterName(api.Namespace, api.Service); err != nil {
+				return err
+			}
+			log.Debug("HTTP registered", "namespace", api.Namespace)
```
Comment (holiman) [inline node/node.go]: Is this really HTTP specific -- or for any protocol?

File: node/rpcstack.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		MaxAge:         600,
+		AllowedHeaders: []string{"*"},
+	})
+	return c.Handler(srv)
+}
+
+// virtualHostHandler is a handler which validates the Host-header of incoming requests.
+// The virtualHostHandler can prevent DNS rebinding attacks, which do not utilize CORS-headers,
+// since they do in-domain requests against the RPC api. Instead, we can see on the Host-header
+// which domain was used, and validate that against a whitelist.
```
Comment (holiman) [inline node/rpcstack.go]: I find the comment a bit confusing. 
```suggestion
// Using virtual hosts can help prevent DNS rebinding attacks, where a 'random' domain name points to 
// the service ip address (but without CORS headers). By verifying the targeted virtual host, we can 
// ensure that it's a destination that the node operator has defined. 
```
And yes I know you only pasted it over, and maybe it was even me who wrote it at some point :)

File: node/node.go [hunk 7]
```diff
@@ -359,22 +363,36 @@ func (n *Node) stopIPC() {
...
 	// Short circuit if the HTTP endpoint isn't being exposed
 	if endpoint == "" {
 		return nil
 	}
-	listener, handler, err := rpc.StartHTTPEndpoint(endpoint, apis, modules, cors, vhosts, timeouts)
+
+	srv := rpc.NewServer()
+	err := RegisterApisFromWhitelist(apis, modules, srv, false)
+	if err != nil {
+		return err // TODO this should return upon failure, right?
```
Comment (renaynay) [inline node/node.go]: this is correct

File: node/node.go (hunk 5 above)
Comment (renaynay) [reply node/node.go:710]: Can it say `RPC registered` instead ?

Comment (fjl) [reply node/node.go:710]: I think we can just remove this message instead.

File: node/node_test.go [hunk 8]
```diff
@@ -597,3 +600,65 @@ func TestAPIGather(t *testing.T) {
...
+		t.Error("could not issue new http request ", err)
+	}
+	wsReq.Header.Set("Connection", "upgrade")
+	wsReq.Header.Set("Upgrade", "websocket")
+	wsReq.Header.Set("Sec-WebSocket-Version", "13")
+	wsReq.Header.Set("Sec-Websocket-Key", "SGVsbG8sIHdvcmxkIQ==")
+
+	wsResponses := make(chan *http.Response)
+	go doHTTPRequest(wsResponses, wsReq, t)
+	wsResponse := <-wsResponses
```
Comment (fjl) [inline node/node_test.go:619]: You're not really using the channel for anything here, consider simply calling doHTTPRequest.

File: node/node_test.go [hunk 9]
```diff
@@ -597,3 +600,65 @@ func TestAPIGather(t *testing.T) {
...
+
+	err = node.startHTTP("127.0.0.1:7453", []rpc.API{}, []string{}, []string{}, []string{}, rpc.HTTPTimeouts{}, []string{})
+	if err != nil {
+		t.Error("could not start http service on node ", err)
+	}
+
+	return node
+}
+
+func doHTTPRequest(responses chan *http.Response, req *http.Request, t *testing.T) {
```
Comment (fjl) [inline node/node_test.go:656]: `*testing.T` should be first parameter.
Please consider removing the channel.

File: cmd/geth/retesteth.go [hunk 10]
```diff
@@ -888,14 +888,22 @@ func retesteth(ctx *cli.Context) error {
...
 	cors := splitAndTrim(ctx.GlobalString(utils.RPCCORSDomainFlag.Name))
 
+	srv := rpc.NewServer()
+
+	err := node.RegisterApisFromWhitelist(rpcAPI, []string{"test", "eth", "debug", "web3"}, srv, false)
+	if err != nil {
+		utils.Fatalf("Could not register RPC apis: %w", err)
+	}
+
+	handler := node.NewHTTPHandlerStack(srv, cors, vhosts)
```
Comment (fjl) [inline cmd/geth/retesteth.go:898]: Please remove blank lines between declarations of `srv`, `err` and `handler`, and add another one after the `handler` declaration. That way creating the handler is its own paragraph.

File: node/rpcstack.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
+// Copyright 2015 The go-ethereum Authors
```
Comment (gballet) [inline node/rpcstack.go:1]: If you create a new file, change the date to 2020.

File: node/rpcstack.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+var gzPool = sync.Pool{
+	New: func() interface{} {
+		w := gzip.NewWriter(ioutil.Discard)
+		return w
+	},
+}
+
+type gzipResponseWriter struct {
```
Comment (gballet) [inline node/rpcstack.go:108]: that code used to be in a separate file, is there a reason for it to be merged here this time instead of `node/gzip.go` ?

Comment (fjl) [reply node/rpcstack.go:108]: I think it's OK to have all of the HTTP handler wrappers in one file. It's not so much code anyway.

**Pull request merged**